### PR TITLE
Adding state field components

### LINKED ISSUE #540: Update ADCS Box Monitor Serializers
There are a lot of fields reported by the ADCS Box that are serialized over I2C in rectangular coordinates and then written to vector state fields in flight software. This includes the wheel speed reads, magnetometer readings, gyroscope readings, and wheel torque command readings. This implication of storing these values in vector state fields, however, is that they are serialized in spherical coordinates and not rectangular coordinates. This creates a loss in precision or wasted bits in our serialization scheme - if you think about inscribing the cube in the sphere or sphere in the cube.

To remedy this, we should take the following action:

* Change the vector serializers for magnetometer, gyroscope, torque, and wheel speed readings to internal state fields.
* Add float serializers for all the above vector components to implement rectangular serializers.

The following precision values were decided upon for the rectangular serializers:

* Wheel speed read: 8 bits per component.
* Magnetometer: 8 bits per component.
* Wheel torque read: 8 bits per component.
* Gyroscope read: 16 bits per component.

All serializer bounds should be pulled from the ADCS constants.

### PR DESCRIPTION
**Pull request by knk38**
# Updating Serializers for ADCS Box Monitor

One thing I wanted to get sure is, since this PR creates a whole bunch of new state fields that would then need to be tested, another solution that Shihao was telling me about was to add a new serializer that would automatically split up data from the vector fields into its components. Then we would just have to call the new serializer on the already defined vector state field.
Fixes #540 
### Summary of changes
- Added a new float serializer for each of the state fields
- Set each new component state field to take data from its respective vector field

### Testing
Since these fields won't be tested by the software unit tests, we may need to change either the testing or this implementation.

### Documentation Evidence
Didnt really need much documentation as this builds from the past state fields

### CODE REVIEW

**Comment by kkrol27** [comment]
Are `.[xyz]` state field name suffixes going to break things?

**Comment by knk38** [comment]
> Are `.[xyz]` state field name suffixes going to break things?

I talked with shihao and the plotter actually automatically combines .[xyz] fields together so it should be compatible, thats why i changed it from the 0,1,2 that i had before

**Comment by shihaocao** [comment]
@knk38 Actually, haha that's not yet a feature, but I think it'd be a good feature to have along with this PR if we did choose to go to the multiple statefield route instead of a specialized rectangular serailizer

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 1]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
     unsigned int offset, Devices::ADCS &_adcs)
     : TimedControlTask<void>(registry, "adcs_monitor", offset),
     adcs_system(_adcs),
-    rwa_speed_rd_sr(adcs::rwa::min_speed_read, adcs::rwa::max_speed_read, 8*3), //full res is 16*3
+    rwa_speed_rd_sr(adcs::rwa::min_speed_read, adcs::rwa::max_speed_read, 16*3), //referenced from I2C_Interface.doc
+    rwa_speed_rd_component_sr(adcs::rwa::min_speed_read,adcs::rwa::max_speed_read, 8),
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:12]: Are we looking for full resolution here? If so, then this needs to be 12 bits and not 8 (read is done with a 12 bit ADC and the I2C register is 16 bits):
```suggestion
    rwa_speed_rd_component_sr(adcs::rwa::min_speed_read,adcs::rwa::max_speed_read, 12),
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 2]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
     rwa_speed_rd_f("adcs_monitor.rwa_speed_rd", rwa_speed_rd_sr),
-    rwa_torque_rd_sr(adcs::rwa::min_torque, adcs::rwa::max_torque, 8*3), //full res is 16*3
+    rwa_speed_rd_x_f("adcs_monitor.rwa_speed_rd.x",rwa_speed_rd_component_sr),
+    rwa_speed_rd_y_f("adcs_monitor.rwa_speed_rd.y",rwa_speed_rd_component_sr),
+    rwa_speed_rd_z_f("adcs_monitor.rwa_speed_rd.z",rwa_speed_rd_component_sr),
+
+    rwa_torque_rd_sr(adcs::rwa::min_torque, adcs::rwa::max_torque, 16*3), //referenced from I2C_Interface.doc
+    rwa_torque_rd_component_sr(adcs::rwa::min_torque,adcs::rwa::max_torque, 8),
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:19]: Same deal here with the torque read component if we want full resolution:
```suggestion
    rwa_torque_rd_component_sr(adcs::rwa::min_torque,adcs::rwa::max_torque, 12),
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 3]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
     rwa_torque_rd_f("adcs_monitor.rwa_torque_rd", rwa_torque_rd_sr),
-    ssa_mode_rd(2),
+    rwa_torque_rd_x_f("adcs_monitor.rwa_torque_rd.x", rwa_torque_rd_component_sr),
+    rwa_torque_rd_y_f("adcs_monitor.rwa_torque_rd.y", rwa_torque_rd_component_sr),
+    rwa_torque_rd_z_f("adcs_monitor.rwa_torque_rd.z", rwa_torque_rd_component_sr),
+
+    ssa_mode_rd(0,2,2), //referenced from Interface.doc
     ssa_mode_f("adcs_monitor.ssa_mode", ssa_mode_rd),
-    ssa_vec_sr(0,1,1 + SerializerConstants::min_fvsz),
+    ssa_vec_sr(-1,1,16*3), //referenced from I2C_Interface.doc
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:27]: The sun sensor vector is actually supposed to be a unit vector. @tanishqaggarwal not 100% sure what a serializer for that looks like but that's what we need here.

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 4]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
     ssa_vec_f("adcs_monitor.ssa_vec", ssa_vec_sr),
-    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 12), // Full resolution to enable scientific analysis of the sun sensing algorithm
+    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 8),
     ssa_voltages_f(),
-    mag1_vec_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 8*3), //full res is 16*3
+
+    mag1_vec_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 16*3), //referenced from I2C_Interface.doc
+    mag1_vec_component_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 8),
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:33]: This needs to be a 16 serializer if we want full resolution:
```suggestion
    mag1_vec_component_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 16),
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 5]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
     mag1_vec_f("adcs_monitor.mag1_vec", mag1_vec_sr),
-    mag2_vec_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 8*3), //full res is 16*3
+    mag1_vec_x_f("adcs_monitor.mag1_vec.x", mag1_vec_component_sr),
+    mag1_vec_y_f("adcs_monitor.mag1_vec.y", mag1_vec_component_sr),
+    mag1_vec_z_f("adcs_monitor.mag1_vec.z", mag1_vec_component_sr),
+
+
+    mag2_vec_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 16*3), //referenced from I2C_Interface.doc
+    mag2_vec_component_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 8),
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:41]: Same thing here if we want full resolution:
```suggestion
    mag2_vec_component_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 16),
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 6]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+
+    gyr_vec_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16*3), //referenced from I2C_Interface.doc
+    gyr_vec_component_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16), 
     gyr_vec_f("adcs_monitor.gyr_vec", gyr_vec_sr),
-    gyr_temp_sr(adcs::imu::min_rd_temp, adcs::imu::max_rd_temp, 8), //full res is 16
+    gyr_vec_x_f("adcs_monitor.gyr_vec.x", gyr_vec_component_sr),
+    gyr_vec_y_f("adcs_monitor.gyr_vec.y", gyr_vec_component_sr),
+    gyr_vec_z_f("adcs_monitor.gyr_vec.z", gyr_vec_component_sr),
+
+    gyr_temp_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 16), //referenced from I2C_Interface.doc
```
Comment (kkrol27) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:54]: Why did we update the bounds for this serializer?

**Review by kkrol27 — COMMENTED** [review]
I think there are some changes that we need to make to this prior to merging. Also, do we need to update telemetry flows or is that going to happen at a later date?

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 7]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
     unsigned int offset, Devices::ADCS &_adcs)
     : TimedControlTask<void>(registry, "adcs_monitor", offset),
     adcs_system(_adcs),
-    rwa_speed_rd_sr(adcs::rwa::min_speed_read, adcs::rwa::max_speed_read, 8*3), //full res is 16*3
+    rwa_speed_rd_sr(adcs::rwa::min_speed_read, adcs::rwa::max_speed_read, 16*3), //referenced from I2C_Interface.doc
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:11]:
```suggestion
    rwa_speed_rd_sr(0,0,1), // Dummy serializer, required for rwa_speed_rd to be writable by psim
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 8]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+    rwa_speed_rd_component_sr(adcs::rwa::min_speed_read,adcs::rwa::max_speed_read, 8),
     rwa_speed_rd_f("adcs_monitor.rwa_speed_rd", rwa_speed_rd_sr),
-    rwa_torque_rd_sr(adcs::rwa::min_torque, adcs::rwa::max_torque, 8*3), //full res is 16*3
+    rwa_speed_rd_x_f("adcs_monitor.rwa_speed_rd.x",rwa_speed_rd_component_sr),
+    rwa_speed_rd_y_f("adcs_monitor.rwa_speed_rd.y",rwa_speed_rd_component_sr),
+    rwa_speed_rd_z_f("adcs_monitor.rwa_speed_rd.z",rwa_speed_rd_component_sr),
+
+    rwa_torque_rd_sr(adcs::rwa::min_torque, adcs::rwa::max_torque, 16*3), //referenced from I2C_Interface.doc
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:18]:
```suggestion
    rwa_torque_rd_sr(adcs::rwa::min_torque, adcs::rwa::max_torque, 16*3), // Dummy serializer, required for rwa_torque_rd to be readable by psim
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp (hunk 3 above)
Comment (tanishqaggarwal) [reply src/fsw/FCCode/ADCSBoxMonitor.cpp:27]: Tracked to fix this on #539

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 9]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
     ssa_mode_f("adcs_monitor.ssa_mode", ssa_mode_rd),
-    ssa_vec_sr(0,1,1 + SerializerConstants::min_fvsz),
+    ssa_vec_sr(-1,1,16*3), //referenced from I2C_Interface.doc
     ssa_vec_f("adcs_monitor.ssa_vec", ssa_vec_sr),
-    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 12), // Full resolution to enable scientific analysis of the sun sensing algorithm
+    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 8),
     ssa_voltages_f(),
-    mag1_vec_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 8*3), //full res is 16*3
+
+    mag1_vec_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 16*3), //referenced from I2C_Interface.doc
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:32]:
```suggestion
    mag1_vec_sr(0, 0, 1), // dummy serializer required for magnetic field vector to be writable from psim
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 10]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+    mag1_vec_component_sr(adcs::imu::min_mag1_rd_mag, adcs::imu::max_mag1_rd_mag, 8),
     mag1_vec_f("adcs_monitor.mag1_vec", mag1_vec_sr),
-    mag2_vec_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 8*3), //full res is 16*3
+    mag1_vec_x_f("adcs_monitor.mag1_vec.x", mag1_vec_component_sr),
+    mag1_vec_y_f("adcs_monitor.mag1_vec.y", mag1_vec_component_sr),
+    mag1_vec_z_f("adcs_monitor.mag1_vec.z", mag1_vec_component_sr),
+
+
+    mag2_vec_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 16*3), //referenced from I2C_Interface.doc
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:40]:
```suggestion
    mag2_vec_sr(0, 0, 1), // dummy serializer required for magnetic field vector to be writable from psim
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 11]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+    mag2_vec_component_sr(adcs::imu::min_mag2_rd_mag, adcs::imu::max_mag2_rd_mag, 8),
     mag2_vec_f("adcs_monitor.mag2_vec", mag2_vec_sr),
-    gyr_vec_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16*3), //full res is 16*3
+    mag2_vec_x_f("adcs_monitor.mag2_vec.x", mag2_vec_component_sr),
+    mag2_vec_y_f("adcs_monitor.mag2_vec.y", mag2_vec_component_sr),
+    mag2_vec_z_f("adcs_monitor.mag2_vec.z", mag2_vec_component_sr),
+
+    gyr_vec_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16*3), //referenced from I2C_Interface.doc
```
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:47]:
```suggestion
    gyr_vec_sr(0, 0, 1), // dummy serializer required for gyroscope reading to be writable from psim
```

**Review by tanishqaggarwal — CHANGES_REQUESTED** [review]

File: src/fsw/FCCode/ADCSBoxMonitor.cpp (hunk 6 above)
Comment (shihaocao) [reply src/fsw/FCCode/ADCSBoxMonitor.cpp:54]: gyr_temp_sr(adcs::imu::min_rd_temp, adcs::imu::max_rd_temp, 8), //full res is 16

is appropriate

Comment (kkrol27) [reply src/fsw/FCCode/ADCSBoxMonitor.cpp:54]:
```suggestion
    gyr_temp_sr(adcs::imu::min_rd_temp, adcs::imu::max_rd_temp, 8),
```

File: src/fsw/FCCode/ADCSBoxMonitor.cpp (hunk 3 above)
Comment (tanishqaggarwal) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:27]:
```suggestion
    ssa_vec_sr(0,1,1), // Dummy serializer, needs to get fixed in #539 
```

Comment (shihaocao) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:27]:
```suggestion
    ssa_vec_sr(0,1,1 + SerializerConstants::min_fvsz),
```

Comment (shihaocao) [reply src/fsw/FCCode/ADCSBoxMonitor.cpp:27]: revert to master

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 12]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+    rwa_torque_rd_y_f("adcs_monitor.rwa_torque_rd.y", rwa_torque_rd_component_sr),
+    rwa_torque_rd_z_f("adcs_monitor.rwa_torque_rd.z", rwa_torque_rd_component_sr),
+
+    ssa_mode_rd(0,2,2), //referenced from Interface.doc
     ssa_mode_f("adcs_monitor.ssa_mode", ssa_mode_rd),
-    ssa_vec_sr(0,1,1 + SerializerConstants::min_fvsz),
+    ssa_vec_sr(-1,1,16*3), //referenced from I2C_Interface.doc
     ssa_vec_f("adcs_monitor.ssa_vec", ssa_vec_sr),
-    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 12), // Full resolution to enable scientific analysis of the sun sensing algorithm
+    ssa_voltage_sr(adcs::ssa::min_voltage_rd, adcs::ssa::max_voltage_rd, 8),
```
Comment (shihaocao) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:29]: unsigned char in pass thru anyway

File: src/fsw/FCCode/ADCSBoxMonitor.cpp (hunk 6 above)
Comment (shihaocao) [reply src/fsw/FCCode/ADCSBoxMonitor.cpp:54]: This shall be restored back to master version with 16 bit res

File: src/fsw/FCCode/ADCSBoxMonitor.cpp [hunk 13]
```diff
@@ -8,24 +8,52 @@ ADCSBoxMonitor::ADCSBoxMonitor(StateFieldRegistry &registry,
...
+
+    gyr_vec_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16*3), //referenced from I2C_Interface.doc
+    gyr_vec_component_sr(adcs::imu::min_rd_omega, adcs::imu::max_rd_omega, 16), 
     gyr_vec_f("adcs_monitor.gyr_vec", gyr_vec_sr),
-    gyr_temp_sr(adcs::imu::min_rd_temp, adcs::imu::max_rd_temp, 8), //full res is 16
+    gyr_vec_x_f("adcs_monitor.gyr_vec.x", gyr_vec_component_sr),
+    gyr_vec_y_f("adcs_monitor.gyr_vec.y", gyr_vec_component_sr),
+    gyr_vec_z_f("adcs_monitor.gyr_vec.z", gyr_vec_component_sr),
+
+    gyr_temp_sr(adcs::imu::min_rd_temp, adcs::imu::max_rd_temp, 16), //full res is 16
```
Comment (shihaocao) [inline src/fsw/FCCode/ADCSBoxMonitor.cpp:54]: bump to full res for consistency

**Review by tanishqaggarwal — APPROVED** [review]

**Review by kkrol27 — APPROVED** [review]

**Pull request merged**